### PR TITLE
Issue #1724: Move logic for version information into Boot classes.

### PR DESCRIPTION
**Pull request by quicksketch**
Fixes https://github.com/drush-ops/drush/issues/1724 for master.

### CODE REVIEW

**Comment by greg-1-anderson** [comment]
This looks good to me.  We are still in the situation where we hope that no code will be calling "drush_drupal_version()", ultimately, but this should help us get along in compatibility mode until then.

This PR does shine a light on an existing problem in Drush:  drush_drupal_version needs to do bootstrappy-like stuff, like loading the autoloader. This could cause problems if some code at some point called drush_drupal_version on more than one site.  To avoid this danger, we should remove the $drupal_root parameter, and force commands that need this facility to instead call the "status" command via drush_invoke_process().  Currently, only the archive commands do this.

I think I'm fine merging this PR as-is, though, and handling the API adjustment in a follow-on issue.

File: includes/drupal.inc [hunk 1]
```diff
@@ -36,29 +36,9 @@ function drush_drupal_version($drupal_root = NULL) {
...
-          if (file_exists($drupal_root . $path)) {
-            require_once $drupal_root . $path;
-            if (defined('VERSION')) {
-              $version = VERSION;
-              break;
-            }
-          }
-        }
+      $bootstrap = drush_bootstrap_class_for_root($drupal_root);
+      if ($bootstrap) {
```
Comment (weitzman) [inline includes/drupal.inc]: We have \Drush\Boot\EmptyBoot class. So should bootstrap class ever be FALSE?

**Comment by weitzman** [comment]
I agree with removing $drupal_root param in a followup. The static caching in this function already makes it useless for calling on two different roots.

Code looks good to me. I added one line note for @greg-1-anderson to review.

**Comment by quicksketch** [comment]
> We are still in the situation where we hope that no code will be calling "drush_drupal_version()", ultimately, but this should help us get along in compatibility mode until then.

Yeah, I thought about converting the existing code to use the Boot classes instead, but for starters I figured it'd be adequate to make it act as a compatibility layer. Maybe we could add a `@deprecated` indicator in the PHPdoc, but I'd defer to you guys on that one.

**Comment by greg-1-anderson** [comment]
So, yes, drush_bootstrap_class_for_root might return NULL, because EmptyBoot always claims that it supports nothing at all.  EmptyBoot is inserted by drush_select_bootstrap_class(). I think the code should be refactored so that EmptyBoot is not added to the candidates, and drush_select_bootstrap_class instantiates one by name, rather than taking the last candidate.

The correct solution, though, is to remove the root parameter from the Drush version functions, and use drush_get_bootstrap_object() instead. Everyone should use that function to get the bootstrap object; only the bootstrap code should use drush_bootstrap_class_for_root() or drush_select_bootstrap_class().  We'd have to fix the archive commands to do this, though.

**Pull request merged**